File: DataStructures/Stacks/atcoder/A51/A51.js
const fs = require('fs');

const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
const Q = parseInt(input[0], 10);

const stack = [];
const output = [];

for (let i = 1; i <= Q; i++) {
    const line = input[i];
    const [type, x] = line.split(' ');

    if (type === '1') {
        // クエリ1: 本を積む
        stack.push(x);
    } else if (type === '2') {
        // クエリ2: 一番上の本を出力
        output.push(stack[stack.length - 1]);
    } else if (type === '3') {
        // クエリ3: 一番上の本を取り除く
        stack.pop();
    }
}

console.log(output.join('\n'));

File: DataStructures/Stacks/atcoder/A51/A51.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	Q, _ := strconv.Atoi(scanner.Text())

	stack := make([]string, 0, Q)
	output := make([]string, 0, Q)

	for i := 0; i < Q; i++ {
		scanner.Scan()
		line := scanner.Text()
		parts := strings.Split(line, " ")

		switch parts[0] {
		case "1":
			stack = append(stack, parts[1])
		case "2":
			output = append(output, stack[len(stack)-1])
		case "3":
			stack = stack[:len(stack)-1]
		}
	}

	fmt.Println(strings.Join(output, "\n"))
}
